enhancements/node-healthcheck-operator.md

```diff
@@ -0,0 +1,435 @@
+putting provider specific annotations.
+Potentially the NHC resource could contain label selectors to track
+repeating remediation and mark them as [incrementally back-off](https://issues.redhat.com/browse/OCPCLOUD-800). e.g
+
+Suggested backoff specification under the NHC resource may look like this:
+```yaml
+kind: NodeHealthCheck
+...
+spec:
+  selector:
+     ...
+  # optional
+  backoff:
+    type: exponential
+    limit: 10m
+```
+
+Caveat of this approach is that we need to take care of conflicting or overlapping backoff definitions.
+Setting global catchall backoff strategy is impossible or just coarse without specifying a selector (how do we know what to backoff from?)
+
+Currently backoff is not implemented.
+
+#### Multiple remediation provider deployed
+
+Multiple providers could act on remediation objects. On non heterogeneous cluster
+this may even make sense. It is not the goal of this controller to coordinate
+those, but providers will need disparate set of nodes to handle based on labels,
+or operator on a node if they can hold a [maintenance lease][node maintenance lease proposal].
+
+
+## Design Details
+
+### Logical Conditions Matching
+
+The resource will be given a list of conditions to track, per a selection of
+nodes, and will evaluate each condition in the list. It's a logical evaluation i.e
+it is enough to evaluate a single condition to true per node to mark as unhealthy.
+
+```
+unhealthyConditions:
+   - type:    "Ready"
+     status:  "Unknown"
+     duration: "300s"
+   - type:    "Ready"
+     status:  "False"
+     duration: "300s"
+```
+In the above example only 1 condition needs to match to mark a node non-healthy.
+
+### Max Unhealthy count
+
+To create remediation object (i.e act for remediation) the overall nodes in the
+selection groups must not exceed `maxUnhealthy` threshold. The value can be
+an int or a percentage, similar to v1/apps `maxUnavailable`:
+```
+spec:
+ maxUnhealthy: 30%
+```
+
+### Default spec values:
+selector: {matchingExpression: { Key: node-role.kubernetes.io/worker, Operator: Exists } }
+maxUnhealthy: 49%
+unhealthyConditions: [ {type: NotReady, status: Unknown, duration: 300s},{ type: Ready, status: False, 300s} ]
+remediationTemplate: n/a
+
+### Mandatory spec fields:
+remediationTemplate
+
+### Optional spec fields:
+  - selector
+  - maxUnhealthy
+  - unhealthyConditions
+
+
+### Status sub-resource fields
+observedNodes - specifies how much nodes are observed by this object selector
+heathyNodes     - specify the number of healthy nodes
+inFlightRemediations - a mapping of a remediaiton object name (== the node name) and the time it was created. Removed
+when the remediation object deleted.
+
+
+### Remediation Object creation
+Add the target Node to ownerReferences. We want that object to be garbage
+collected with the referenced Node
+
+
+### Controller details
+leader election: true
+Sync period: 60s - determines how frequent reconciles happen regardless of resource changes.
```

Comment:
Can the user change this value?

Reply:
Not normally, and 60s is far too frequent IMO. This means the controller will be doing a full list against the API for all objects it cares about every 60s. Typically this would be more on the order of minutes than seconds, eg IIRC, in MAPI we use 10m for the sync period. Others even use hours.

I think having a sync period of 60s puts unnecessary load on the API server without much benefit